File: make_minmax.r
# convert daily weather into min/max weather format
library(data.table)
source("util.r")

# This function is for the method that takes old traditional 1pm weather (labelled here as noon)
# and estimates a diurnal temperature range to fit into the scheme that flows into the max/min method
# Temp input in Celsius   RH input in Percent.   These should be that traditional 1pm values
# Written as a function to enable upgrading later if needs be
temp_min_max <- function(temp_noon, rh_noon) {
  # FIX: verify what this should be if temp_noon is negative
  temp_range <- 17 - 0.16 * rh_noon + 0.22 * temp_noon
  temp_max <- ifelse(((temp_noon < 3) & (rh_noon == 100)) | (temp_range < 2),
    temp_noon + (temp_range / 2.0),
    temp_noon + 2
  )
  temp_min <- ifelse(((temp_noon < 3) & (rh_noon == 100)) | (temp_range < 2),
    temp_noon - (temp_range / 2.0),
    temp_max - temp_range
  )
  # HACK: for now just sort so we know it's min, max
  return(list(pmin(temp_min, temp_max), pmax(temp_min, temp_max)))
}
#' Convert daily noon values stream to daily min/max values stream.
#' Uses values from statistics to do the conversion.
#'
#' @param df        daily noon values weather stream [lat, long, yr, mon, day, temp, rh, ws, prec]
#' @return          daily min/max values weather stream [lat, long, yr, mon, day, temp_min, temp_max, rh_min, rh_max, ws_min, ws_max, prec]
#' @export daily_to_minmax
daily_to_minmax <- function(df) {
  df <- data.table(df)
  hadId <- FALSE
  if ("id" %in% tolower(colnames(df))) {
    hadId <- TRUE
  }
  df[, c("temp_min", "temp_max") := temp_min_max(temp, rh)]
  df[, q := findQ(temp, rh)]
  df[, rh_min := pmin(100, pmax(0, findrh(q, temp_max)))]
  df[, rh_max := pmin(100, pmax(0, findrh(q, temp_min)))]
  df[, ws_min := 0.15 * ws]
  df[, ws_max := 1.25 * ws]
  if (hadId) {
    df <- df[, c("id", "lat", "long", "yr", "mon", "day", "temp_min", "temp_max", "rh_min", "rh_max", "ws_min", "ws_max", "prec")]
  } else {
    df <- df[, c("lat", "long", "yr", "mon", "day", "temp_min", "temp_max", "rh_min", "rh_max", "ws_min", "ws_max", "prec")]
  }
  return(df)
}

# so this can be run via Rscript
if ("--args" %in% commandArgs()) {
  args <- commandArgs(trailingOnly = TRUE)
  if (2 == length(args)) {
    inp <- args[1]
    out <- args[2]
    df <- as.data.table(read.csv(inp))
    df_minmax <- daily_to_minmax(df)
    save_csv(df_minmax, out)
  } else {
    message("Wrong number of arguments")
  }
}

File: make_minmax.c
/*
inputs daily weather stream
outputs min/max weather stream
*/

#include <stdio.h>
#include <math.h>
#include <float.h>
#include <stdlib.h>
#include <string.h>
#include "util.h"

/*  This function is for the method that takes old traditional 1pm weather (labelled here as noon)
    and estimates a diurnal temperature range to fit into the scheme that flows into the max/min method
    Temp input in Celsius   RH input in Percent.   These should be that traditional 1pm values
     Written as a function to enable upgrading later if needs be
*/
void temp_min_max(double temp_noon, double rh_noon, double *temp_min, double *temp_max)
{
  double temp_range = 17 - 0.16 * rh_noon + 0.22 * temp_noon;
  if (temp_range <= 2)
  {
    *temp_max = temp_noon + 1;
    *temp_min = temp_noon - 1;
  }
  else
  {
    *temp_max = temp_noon + 2;
    *temp_min = *temp_max - temp_range;
  }
}

int main(int argc, char *argv[])
{
  /*  CSV headers */
  static const char *header = "lat,long,yr,mon,day,temp,rh,ws,prec";
  static const char *header_out = "lat,long,yr,mon,day,temp_min,temp_max,rh_min,rh_max,ws_min,ws_max,prec";
  if (argc != 3)
  {
    printf("Command line:   %s <input file> <output file>\n\n", argv[0]);
    printf("INPUT FILE format must be DAILY weather data, comma seperated and take the form\n");
    printf("%s\n\n", header);
    printf("All times should be local standard time\n");
    exit(1);
  }
  printf("Opening input file >>> %s   \n", argv[1]);
  FILE *inp = fopen(argv[1], "r");
  if (NULL == inp)
  {
    printf("\n\n ***** FILE  %s  does not exist\n", argv[1]);
    exit(1);
  }
  check_header(inp, header);
  FILE *out = fopen(argv[2], "w");
  fprintf(out, "%s\n", header_out);
  struct row_daily cur;
  int err = read_row_daily(inp, &cur);
  while (err > 0)
  {
    double temp_min;
    double temp_max;
    temp_min_max(cur.temp, cur.rh, &temp_min, &temp_max);
    double q = findQ(cur.temp, cur.rh);
    double rh_min = _min(100, _max(0, findrh(q, temp_max)));
    double rh_max = _min(100, _max(0, findrh(q, temp_min)));
    double ws_min = 0.15 * cur.wind;
    double ws_max = 1.25 * cur.wind;
    save_csv(out,
             "%.4f,%.4f,%4d,%02d,%02d,%.1f,%.1f,%.0f,%.0f,%.1f,%.1f,%.2f\n",
             cur.lat,
             cur.lon,
             cur.year,
             cur.mon,
             cur.day,
             temp_min,
             temp_max,
             rh_min,
             rh_max,
             ws_min,
             ws_max,
             cur.rain);
    printf("%.4f,%.4f,%4d,%02d,%02d,%.1f,%.1f,%.0f,%.0f,%.1f,%.1f,%.2f\n",
           cur.lat,
           cur.lon,
           cur.year,
           cur.mon,
           cur.day,
           temp_min,
           temp_max,
           rh_min,
           rh_max,
           ws_min,
           ws_max,
           cur.rain);
    err = read_row_daily(inp, &cur);
  } /* end the main while(err>0)  */
  fclose(inp);
  fclose(out);
  return 0;
}
